### PR TITLE
Add Bedrock contract tests.

### PR DESCRIPTION
**Pull request by mxiamxia**
The PR is a follow up on bedrock service support PR:
https://github.com/aws-observability/aws-otel-java-instrumentation/pull/826

We add contract tests for following Bedrock services that covers all resource attributes we newly support:
1. Bedrock API: `GetGuardrail` 
2. BedrockAgent APIs: `GetAgent`, `GetDataSource`, `GetKnowledgeBase`
3. BedrockRuntime API: `InvokeModel`
4. BedrockAgentRuntime API: `InvokeAgent`

Upgrade `aws-java-sdk` V1 and V2 to the latest version so that to support Bedrock services API calls.
Upgrade `localstack/localstack` image to the latest version `3.5.0` so resolve the SQS API call issue using `localstack/localstack:2.0.1`: https://github.com/localstack/localstack/issues/9610
Add `.withEnv("LOCALSTACK_HOST", "127.0.0.1")` for `LocalStackContainer` to workaround the 3.X version limitation: https://github.com/localstack/localstack/issues/9753

**Contract test limitation:**
The contract tests in current repo is using [LocalStackContainer](https://github.com/aws-observability/aws-otel-java-instrumentation/blob/e5428dde9c8059d39782044b39091afca28ddf0e/appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/awssdk/base/AwsSdkBaseTest.java#L39) to serve AWS SDK service calls. But it doesn't has bedrock related service support (This is [the full service list](https://github.com/testcontainers/testcontainers-java/blob/1f38f0d9604edb9e89fd3b3ee1eff6728e2d1e07/modules/localstack/src/main/java/org/testcontainers/containers/localstack/LocalStackContainer.java#L402
) it support.). In this case, no matter which bedrock API we call in contract test, the response will always be 4XX.

As a workaround, we We point all Bedrock related request endpoints to the local app and then specifically handle each request to return the expected response to make sure we receive http response with expected status code and attributes. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: appsignals-tests/images/aws-sdk/aws-sdk-v2/src/main/java/com/amazon/sampleapp/App.java [hunk 1]
```diff
@@ -509,7 +530,262 @@ private static void setupS3() {
...
+            .endpointOverride(URI.create("http://bedrock.test:8080"))
+            .region(Region.US_WEST_2)
+            .credentialsProvider(CREDENTIALS_PROVIDER)
+            .build();
+
+    // Response with GetKnowledgeBaseRequest.
+    get(
+        "/knowledgebases/:knowledgeBaseId",
+        (req, res) -> {
+          String knowledgeBaseId = req.params(":knowledgeBaseId");
```
Comment (thpierce) [inline appsignals-tests/images/aws-sdk/aws-sdk-v2/src/main/java/com/amazon/sampleapp/App.java:570]: Is this output identical to aws-sdk-v1 response? Can we deduplicate and have some shared code? Just a thought.

File: appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/awssdk/base/AwsSdkBaseTest.java [hunk 2]
```diff
@@ -35,16 +38,26 @@
...
 public abstract class AwsSdkBaseTest extends ContractTestBase {
+  protected static final Logger LOGGER = Logger.getLogger(AwsSdkBaseTest.class.getName());
+
+  static {
+    ConsoleHandler handler = new ConsoleHandler();
+    handler.setLevel(Level.INFO);
+    LOGGER.addHandler(handler);
+    LOGGER.setLevel(Level.INFO);
+    LOGGER.setUseParentHandlers(false);
+  }
```
Comment (thpierce) [inline appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/awssdk/base/AwsSdkBaseTest.java:49]: Why is this needed

**Comment by thpierce** [comment]
Is the overview up to date? It mentions changes that I don't see, like `inject_200_success`, `inject_500_error`, and `_assert_semantic_conventions_span_attributes`.

File: appsignals-tests/images/aws-sdk/aws-sdk-v2/build.gradle.kts [hunk 3]
```diff
@@ -27,13 +27,18 @@ dependencies {
...
-  implementation(enforcedPlatform("software.amazon.awssdk:bom:2.20.102"))
+  implementation(enforcedPlatform("software.amazon.awssdk:bom:2.26.20"))
   implementation("software.amazon.awssdk:s3")
   implementation("software.amazon.awssdk:dynamodb")
   implementation("software.amazon.awssdk:sqs")
   implementation("software.amazon.awssdk:kinesis")
+  implementation("software.amazon.awssdk:bedrock:2.26.20")
+  implementation("software.amazon.awssdk:bedrockagent:2.26.20")
+  implementation("software.amazon.awssdk:bedrockruntime:2.26.20")
+  implementation("software.amazon.awssdk:bedrockagentruntime:2.26.20")
```
[the remark this answers is not in the document]
Comment (zzhlogin) [reply appsignals-tests/images/aws-sdk/aws-sdk-v2/build.gradle.kts:38]: Removed versions

File: appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/awssdk/v1/AwsSdkV1Test.java [hunk 4]
```diff
@@ -165,4 +200,34 @@ void testKinsesisError() throws Exception {
   void testKinesisFault() throws Exception {
     doTestKinesisFault();
   }
+
+  @Test
+  void testBedrockGetKnowlesgeBase() {
+    doTestBedrockGetKnowlesgeBase();
```
[the remark this answers is not in the document]
Comment (zzhlogin) [reply appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/awssdk/v1/AwsSdkV1Test.java:206]: Updated.

File: appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/utils/SemanticConventionsConstants.java [hunk 5]
```diff
@@ -58,6 +58,11 @@ public class SemanticConventionsConstants {
   public static final String AWS_QUEUE_URL = "aws.queue.url";
   public static final String AWS_QUEUE_NAME = "aws.queue.name";
   public static final String AWS_STREAM_NAME = "aws.stream.name";
+  public static final String AWS_KNOWLEDGEBASE_ID = "aws.bedrock.knowledge_base.id";
+  public static final String AWS_DATASOURCE_ID = "aws.bedrock.data_source.id";
```
[the remark this answers is not in the document]
Comment (zzhlogin) [reply appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/utils/SemanticConventionsConstants.java:62]: Updated

File: appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/utils/SemanticConventionsConstants.java [hunk 6]
```diff
@@ -58,6 +58,11 @@ public class SemanticConventionsConstants {
...
+  public static final String AWS_AGENT_ID = "aws.bedrock.agent.id";
+  public static final String AWS_GUARDRAIL_ID = "aws.bedrock.guardrail.id";
+  public static final String AWS_BEDROCK_RUNTIME_MODEL_ID = "gen_ai.request.model";
```
[the remark this answers is not in the document]
Comment (zzhlogin) [reply appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/utils/SemanticConventionsConstants.java:65]: Updated

File: appsignals-tests/images/aws-sdk/aws-sdk-v1/src/main/java/com/amazon/sampleapp/App.java [hunk 7]
```diff
@@ -206,7 +224,7 @@ private static void setupSqs() {
           try {
             errorClient.sendMessage(sendMessageRequest);
           } catch (Exception ex) {
-
+            ex.printStackTrace();
```
[the remark this answers is not in the document]
Comment (zzhlogin) [reply appsignals-tests/images/aws-sdk/aws-sdk-v1/src/main/java/com/amazon/sampleapp/App.java:227]: Removed try-catch here as we don't expect an exception. The exception is only expected for error & fault test cases.

File: appsignals-tests/images/aws-sdk/aws-sdk-v2/src/main/java/com/amazon/sampleapp/App.java (hunk 1 above)
Comment (zzhlogin) [reply appsignals-tests/images/aws-sdk/aws-sdk-v2/src/main/java/com/amazon/sampleapp/App.java:570]: Yes, deduplicated it and move the share code into a shared module in `aws-sdk-base/.../Utils.java`

File: appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/awssdk/base/AwsSdkBaseTest.java (hunk 2 above)
Comment (zzhlogin) [reply appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/awssdk/base/AwsSdkBaseTest.java:49]: This was used for debugging purpose. Removed it.

File: appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/awssdk/base/AwsSdkBaseTest.java [hunk 8]
```diff
@@ -1559,4 +1621,379 @@ protected void doTestKinesisFault() throws Exception {
...
+        localService,
+        localOperation,
+        getBedrockAgentServiceName(),
+        "GetKnowledgeBase",
+        type,
+        identifier,
+        0.0);
+  }
+
+  protected void doTestBedrockFault() {
```
[the remark this answers is not in the document]
Comment (zzhlogin) [reply appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/awssdk/base/AwsSdkBaseTest.java]: As discussed offline, we decided to reduce the 6 cases in Agents for Amazon Bedrock down to 3. Update the PR to have following tests:
`testBedrockAgentAgentId` -> 1.i 1.ii
`testBedrockAgentGetKnowledgeBaseIdFromRequest` -> 1.iii, 1.iv
`testBedrockAgentDataSourceId` -> 1.v, 1.vi
`testBedrockAgentRuntimeAgentId` -> 2.i, 
`testBedrockAgentRuntimeKnowledgeBaseId` -> 2.ii
`testBedrockGuardrailId` -> 3.i
`testBedrockRuntimeModelId` -> 4.i

For AWS SDK V2, a gap is identified that: the knowledgeBaseId is not extracted from bedrockAgentRuntime APIs. We will address this gap along with the other upcoming AWS Resources implementation PRs. In this case, there will be no contract test for 2.ii for SDK V2.

**Comment by zzhlogin** [comment]
> Is the overview up to date? It mentions changes that I don't see, like `inject_200_success`, `inject_500_error`, and `_assert_semantic_conventions_span_attributes`.

Seems like this is a rebased PR, and the PR description was not copied correctly. Updated the description from [original PR](https://github.com/aws-observability/aws-otel-java-instrumentation/pull/848).

**Review by mxiamxia — COMMENTED** [review]
the updates LGTM! ty!

File: appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/awssdk/v1/AwsSdkV1Test.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (thpierce) [reply appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/awssdk/v1/AwsSdkV1Test.java]: Not blocking: Missed testBedrockAgentGetKnowlesgeBaseIdFromRequest

File: appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/awssdk/base/AwsSdkBaseTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (thpierce) [reply appsignals-tests/contract-tests/src/test/java/software/amazon/opentelemetry/appsignals/test/awssdk/base/AwsSdkBaseTest.java]: LGTM

File: appsignals-tests/images/aws-sdk/aws-sdk-base/src/main/java/com/amazon/sampleapp/Utils.java [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
```
Comment (thpierce) [inline appsignals-tests/images/aws-sdk/aws-sdk-base/src/main/java/com/amazon/sampleapp/Utils.java:14]: Not blocking: We can use the short form:

```
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
```

**Review by thpierce — APPROVED** [review]

**Review by thpierce — APPROVED** [review]

**Review by vasireddy99 — APPROVED** [review]
